src/v/transform/transform_manager.cc

```diff
@@ -0,0 +1,386 @@
+    // the backoff policy for this processor for when we attempt to restart
+    // the processor. If it's been enough time since our last restart of the
+    // processor we will reset this.
+    rpc::backoff_policy _backoff
+      = rpc::make_exponential_backoff_policy<ss::lowres_clock>(
+        base_duration, max_duration);
+};
+
+} // namespace
+
+// The underlying container for holding processors and indexing them correctly.
+//
+// It's possible for there to be the same ntp with multiple transforms and it's
+// also possible for a transform to be attached to multiple ntps, so the true
+// "key" for a processor is the tuple of (id, ntp), hence the need for the
+// complicated table of indexes.
+class processor_table {
+    struct key_index_t {};
+    struct ntp_index_t {};
+    struct id_index_t {};
+
+    class entry_t {
+    public:
+        entry_t(
+          std::unique_ptr<transform::processor> processor,
+          ss::lw_shared_ptr<transform::probe> probe)
+          : _processor(std::move(processor))
+          , _probe(std::move(probe))
+          , _backoff(std::make_unique<processor_backoff>()) {}
+
+        processor* processor() const { return _processor.get(); }
+        ss::lw_shared_ptr<probe> probe() const { return _probe; }
+        processor_backoff* backoff() const { return _backoff.get(); }
+
+        const model::ntp& ntp() const { return _processor->ntp(); }
+        model::transform_id id() const { return _processor->id(); }
+
+    private:
+        std::unique_ptr<transform::processor> _processor;
+        ss::lw_shared_ptr<transform::probe> _probe;
+        std::unique_ptr<processor_backoff> _backoff;
+    };
+
+    using underlying_t = boost::multi_index::multi_index_container<
+      entry_t,
+      boost::multi_index::indexed_by<
+        // uniquely indexed by id and ntp together
+        boost::multi_index::hashed_unique<
+          boost::multi_index::tag<key_index_t>,
+          boost::multi_index::composite_key<
+            entry_t,
+            boost::multi_index::
+              const_mem_fun<entry_t, model::transform_id, &entry_t::id>,
+            boost::multi_index::
+              const_mem_fun<entry_t, const model::ntp&, &entry_t::ntp>>,
+          boost::multi_index::composite_key_hash<
+            std::hash<model::transform_id>,
+            std::hash<model::ntp>>>,
+        // indexed by ntp
+        boost::multi_index::hashed_non_unique<
+          boost::multi_index::tag<ntp_index_t>,
+          boost::multi_index::
+            const_mem_fun<entry_t, const model::ntp&, &entry_t::ntp>,
+          std::hash<model::ntp>,
+          std::equal_to<>>,
+        // indexed by id
+        boost::multi_index::hashed_non_unique<
+          boost::multi_index::tag<id_index_t>,
+          boost::multi_index::
+            const_mem_fun<entry_t, model::transform_id, &entry_t::id>,
+          std::hash<model::transform_id>,
+          std::equal_to<>>>>;
```

Comment:
lol what?

maybe a comment above explaining the data model here? also do we know that there aren't any sneaky vectors or other contiguous memory data structures constructed in this data structure that might grow large and bite us?

Reply:
lol yeah sort of massive, I don't love it. It is probably better to do the manual checks ourselves, I'll remove the usage of multi_index because I'm not sure there aren't sneaky vectors at all.

Reply:
I removed it for understandable memory usage.

Reply:
>I removed it for understandable memory usage.

bummer, it was a really impressive type definition